runtime/runtime/src/metrics.rs

```diff
@@ -105,3 +106,196 @@ pub static FUNCTION_CALL_PROCESSED_CACHE_ERRORS: Lazy<IntCounterVec> = Lazy::new
+    )
+    .unwrap()
+});
+static CHUNK_INC_RECEIPTS_COMPUTE_HIST: Lazy<HistogramVec> = Lazy::new(|| {
+    try_create_histogram_vec(
+        "near_chunk_inc_receipt_compute_hist",
+        "Compute time for applying incoming receipts by chunk, as a histogram in ms",
+        &["shard_id"],
+        buckets_for_compute(),
+    )
+    .unwrap()
+});
+static CHUNK_INC_RECEIPTS_TGAS_HIST: Lazy<HistogramVec> = Lazy::new(|| {
+    try_create_histogram_vec(
+        "near_chunk_inc_receipt_tgas_hist",
+        "Tgas burnt for applying incoming receipts by chunk, as a histogram in ms",
+        &["shard_id"],
+        buckets_for_gas(),
+    )
+    .unwrap()
+});
+static CHUNK_TX_COMPUTE_HIST: Lazy<HistogramVec> = Lazy::new(|| {
+    try_create_histogram_vec(
+        "near_chunk_tx_compute_hist",
+        "Compute time for transaction validation by chunk, as a histogram in ms",
+        &["shard_id"],
+        Some(vec![50., 100., 200., 300., 400., 500., 600.0]),
+    )
+    .unwrap()
+});
+static CHUNK_TX_TGAS_HIST: Lazy<HistogramVec> = Lazy::new(|| {
+    try_create_histogram_vec(
+        "near_chunk_tx_tgas_hist",
+        "Tgas burnt for transaction validation by chunk, as a histogram",
+        &["shard_id"],
+        Some(vec![50., 100., 200., 300., 400., 500.]),
+    )
+    .unwrap()
+});
+
+/// Buckets used for burned gas in receipts.
+///
+/// The maximum possible is 1300 Tgas for a full chunk.
+/// But due to the split between types of receipts, it should be quite rare to
+/// see more than 1000.
+fn buckets_for_gas() -> Option<Vec<f64>> {
+    Some(vec![50., 100., 200., 300., 400., 500., 600., 700., 800., 900., 1000., 1100., 1200.])
+}
+/// Buckets used for receipt compute time usage, in ms.
+///
+/// Ideally the range should be 0-1300 ms. But when we increase te compute cost
+/// compared to gas cost, it can be higher than that.
+///
+/// Nevertheless, it should be on the rare side to have > 1300 ms and hence it
+/// is not worth collecting at a high granularity above that. Here we pick one
+/// bucket split at 2000 ms to easily single out heavy undercharging.
+fn buckets_for_compute() -> Option<Vec<f64>> {
+    Some(vec![
+        50., 100., 200., 300., 400., 500., 600., 700., 800., 900., 1000., 1100., 1200., 1300.,
+        2000.,
+    ])
+}
+
+/// Helper struct to collect partial costs of `Runtime::apply` and reporting it
+/// atomically.
+///
+/// Note: All numbers below are cumulative, to get partial costs they need to be
+/// subtracted from each other as one in `fn report()`.
+#[derive(Debug, Default)]
+pub struct ApplyMetrics {
+    tx_compute_usage: u64,
+    tx_gas: u64,
+    local_receipts_compute_usage: u64,
+    local_receipts_gas: u64,
+    delayed_receipts_compute_usage: u64,
+    delayed_receipts_gas: u64,
+    incoming_receipts_compute_usage: u64,
+    incoming_receipts_gas: u64,
+}
+
+impl ApplyMetrics {
+    pub fn tx_processing_done(&mut self, accumulated_gas: u64, accumulated_compute: u64) {
```

Comment:
Idea: An alternative way to rewrite this would be to do the subtractions here and simplify the reporting code. This might be a bit simpler to read as the code would be more uniform:

```rust
fn update_accumulated(&mut self, gas: u64, compute: u64) -> (u64, u64) {
  let delta = (self.accumulated_gas.saturating_sub(gas), self.accumulated_compute.saturating_sub(compute));
  self.accumulated_gas = gas;
  self.accumulated_compute = compute;
  delta
}

pub fn local_receipts_done(&mut self, accumulated_gas: u64, accumulated_compute: u64) {
  self.local_receipts_gas, self.local_receipts_compute_usage = self.update_accumulated(accumulated_gas, accumulated_compute);
}
```

Reply:
this is so much better, thanks for the idea!